sycamore/transforms/schema.py

```diff
@@ -0,0 +1,180 @@
+from abc import ABC, abstractmethod
+from typing import Callable, Any, Optional
+import json
+import re
+
+from ray.data import Dataset
+
+from sycamore.data import Element, Document
+from sycamore.plan_nodes import Node, Transform
+from sycamore.llms import LLM
+from sycamore.transforms.map import generate_map_function
+from sycamore.llms.prompts import (
+    SCHEMA_ZERO_SHOT_GUIDANCE_PROMPT,
+    SCHEMA_ZERO_SHOT_GUIDANCE_PROMPT_CHAT,
+    PROPERTIES_ZERO_SHOT_GUIDANCE_PROMPT,
+    PROPERTIES_ZERO_SHOT_GUIDANCE_PROMPT_CHAT,
+)
+
+
+def element_list_formatter(elements: list[Element]) -> str:
+    query = ""
+    for i in range(len(elements)):
+        query += f"ELEMENT {i + 1}: {elements[i].text_representation}\n"
+    return query
+
+
+class SchemaExtractor(ABC):
+    def __init__(self, entity_name: str):
+        self._entity_name = entity_name
+
+    @abstractmethod
+    def extract_schema(self, document: Document) -> Document:
+        pass
+
+class PropertyExtractor(ABC):
+    def __init__(self,):# properties: list[str]):
+       # self._properties = properties
+       pass
+
+    @abstractmethod
+    def extract_properties(self, document: Document) -> Document:
+        pass
+
+
+class OpenAISchema(SchemaExtractor):
+    """
+    OpenAISchema uses one of OpenAI's language model (LLM) for schema extraction.
+    """
+
+    def __init__(
+        self,
+        entity_name: str,
+        llm: LLM,
+        num_of_elements: int = 35,
+        prompt_formatter: Callable[[list[Element]], str] = element_list_formatter,
+    ):
+        super().__init__(entity_name)
+        self._llm = llm
+        self._num_of_elements = num_of_elements
+        self._prompt_formatter = prompt_formatter
+
+    def extract_schema(self, document: Document) -> Document:
+        entities = self._handle_zero_shot_prompting(document)
+
+        try:
+            payload = entities["answer"]
+            pattern = r'```json([\s\S]*?)```'
+            match = re.match(pattern, payload).group(1)
```

Comment:
you might want to look at https://www.npmjs.com/package/jsonrepair, I think it'll handle a few more formatting issues. If you don't see an issue that's fine

Reply:
At the moment not having issues with malformed JSON but we can revisit if testing reveals problems with this.